### DISCUSSION TITLE
Question regarding `auto_session_allowed_dirs`

### DISCUSSION THREAD

**Discussion by MijikHna** [rmagatti/auto-session#350] [Q&A] [answered]
After reading [this discussion](https://github.com/rmagatti/auto-session/discussions/254) I have removed `auto_session_suppress_dirs` and just wanted to use `auto_session_allowed_dirs` Here is my config
```lua
auto_session.setup({
    ...
    auto_session_allowed_dirs = {
        "~/.dotfiles",
        "/media/" .. os.getenv("ME") .. "/WindowsD/Coding/*"
    },
    ...
})
```
But neither the `.dotfiles` session nor sessions for subdirectories of `Coding` directory will be saved and restored. I also tried just `"/media/" .. os.getenv("ME") .. "/WindowsD/Coding`.
Maybe I just misunderstood those both configuration settings?

**Comment by cameronr** [comment]
Hmm, that should be how it works. Can you:

1. post your whole auto-session config

2. add `log_level = 'debug'` to your config and then post what `:messages` or `:Notifications` shows?

3. I assume you're on windows. What does `os.getenv("ME")` return?

**Comment by MijikHna** [comment]
Here is my full configuration
```lua
return {
  "rmagatti/auto-session",
  config = function()
    local auto_session = require("auto-session")

    vim.o.sessionoptions = "blank,buffers,curdir,folds,tabpages,winsize,winpos,terminal,localoptions"

    auto_session.setup({
      log_level                 = "info",
      auto_save_enabled         = true,
      auto_restore_enabled      = true,
      auto_session_allowed_dirs = {
        "~/.dotfiles",
        "/media/" .. os.getenv("ME") .. "/WindowsD/Coding/*"
      },
      cwd_change_handling       = {
        restore_upcoming_session = true,
      }
    })
  end
}
```

I am actually on Linux, but I keep some of my code on Windows Partition to access it also on Windows.
os.getenv("ME"), actually is my username on Linux.

Here is my log for an directory in *Coding* folder
```txt
2024-08-14T11:42:12  DEBUG auto-session DEBUG: Lazy is loaded, but not visible, will try to restore session
2024-08-14T11:42:12  DEBUG auto-session DEBUG: enabled_for_command_line_argv, launch_argv: {}
2024-08-14T11:42:12  DEBUG auto-session DEBUG: No arguments, saving/restoring enabled
2024-08-14T11:42:12  DEBUG auto-session DEBUG: find_matching_directory {
  dirToFind = "/media/kirill/WindowsD/Coding/Projects/PLAYAREA-Spotify/vue",
  dirs = {}
}
2024-08-14T11:42:12  DEBUG auto-session DEBUG: suppress_session didn't find a match, returning false
2024-08-14T11:42:12  DEBUG auto-session DEBUG: RestoreSessionFromDir start { "/home/kirill/.local/share/nvim/sessions/" }
2024-08-14T11:42:12  DEBUG auto-session DEBUG: RestoreSessionFromDir validated session_dir:  /home/kirill/.local/share/nvim/sessions/
2024-08-14T11:42:12  DEBUG auto-session DEBUG: get_session_file_name no session_name, using cwd: /media/kirill/WindowsD/Coding/Projects/PLAYAREA-Spotify/vue
2024-08-14T11:42:12  DEBUG auto-session DEBUG: RestoreSessionFromDir escaped session name: %2Fmedia%2Fkirill%2FWindowsD%2FCoding%2FProjects%2FPLAYAREA-Spotify%2Fvue.vim
2024-08-14T11:42:12  DEBUG auto-session DEBUG: RestoreSessionFile restoring session from: /home/kirill/.local/share/nvim/sessions/%2Fmedia%2Fkirill%2FWindowsD%2FCoding%2FProjects%2FPLAYAREA-Spotify%2Fvue.vim
2024-08-14T11:42:12  DEBUG auto-session DEBUG: Config at start of setup {
  conf = {
    args_allow_files_auto_save = false,
    args_allow_single_directory = true,
    auto_restore_enabled = true,
    auto_restore_lazy_delay_enabled = true,
    auto_save_enabled = true,
    auto_session_allowed_dirs = { "~/.dotfiles", "/media/kirill/WindowsD/Coding/*" },
    auto_session_create_enabled = true,
    auto_session_enable_last_session = false,
    auto_session_enabled = true,
    auto_session_root_dir = "/home/kirill/.local/share/nvim/sessions/",
    auto_session_use_git_branch = false,
    close_unsupported_windows = true,
    cwd_change_handling = {
      restore_upcoming_session = true
    },
    log_level = "debug",
    session_lens = {
      buftypes_to_ignore = {},
      load_on_setup = true,
      mappings = {
        alternate_session = { "i", "<C-S>" },
        delete_session = { "i", "<C-D>" }
      },
      session_control = {
        control_dir = "/home/kirill/.local/share/nvim/auto_session/",
        control_filename = "session_control.json"
      }
    },
    silent_restore = true
  }
}
2024-08-14T11:42:12  DEBUG auto-session DEBUG: Root dir set to: /home/kirill/.local/share/nvim/sessions/
2024-08-14T11:42:12  DEBUG auto-session DEBUG: Loading session lens on setup
2024-08-14T11:42:12  ERROR auto-session ERROR: Error restoring session, disabling auto save.
Set silent_restore = false in the config for a more detailed error message.
Error: E116: Invalid arguments for function dictwatcherdel
2024-08-14T11:42:12  DEBUG auto-session DEBUG: No session restored, call no_restore hooks
```

**Reply by cameronr** [reply]
Thanks for all of that info. It looks like there's an error while restoring the session, specifically with the `dictwatcherdel` command. I don't normally see that in session files. Can you post the contents of your session file so I can see more context:

`/home/kirill/.local/share/nvim/sessions/%2Fmedia%2Fkirill%2FWindowsD%2FCoding%2FProjects%2FPLAYAREA-Spotify%2Fvue.vim`

Also, I want to make sure session saving is working as expected. Can you open one of your projects and then run this:

`:lua require('auto-session').AutoSaveSession()`

That will run the autosave logic but you'll still be able to see the log. Please post the log here so I can make sure auto-saving is working correctly.

**Reply by MijikHna** [reply]
`:lua require('auto-session').AutoSaveSession()` give me this error message

```txt
2024-08-19T16:38:15  DEBUG auto-session DEBUG: launch_argv is nil, saving/restoring enabled
2024-08-19T16:38:15  DEBUG auto-session DEBUG: auto_save_conditions_met: auto_save() is false, returning false
2024-08-19T16:38:15  DEBUG auto-session DEBUG: Auto save conditions not met
```
Maybe also worth mentioning `vue` project is in following location `/media/kirill/WindowsD/Coding/Projects/PLAYAREA-Spotify/vue`. Maybe one `*` in my config not fully correct?


Maybe I just should remove all the files from `~/.local/share/nvim/autosession`, `~/.local/share/nvim/sesssions` and start from scratch. Maybe in my previous configuration was an error and wrong input has been written into session file? I mean you said my config looks actually good.

**Reply by cameronr** [reply]
Auto-saving is disabled when there's an error loading the session so that's why it's not saving a session when you exit.

Before just deleting all of your session data, can you try either manually saving a session after it's loaded via `:SessionSave` or you can re-enable auto-saving after by doing `:SessionToggleAutoSave`. It should give you a notification that auto-saving is re-enabled. Then exit and see if it loads that new session correctly.

If it doesn't, can you add `silent_restore = false` to your auto-session config and then post both the log of the session loading failure and the full session file (the session file above doesn't have the error message from the earlier log so trying to see them together).

thanks for all of the debugging help!

**Reply by MijikHna** [reply]
`:SessionSave` works, but disabling and reenabling `AutoSave` doens't work. Only my manually saved *Session* will  be restored. 
I start to think the problem is ` "/media/" .. os.getenv("ME") .. "/WindowsD/Coding/*"` line. I assume **auto-session** considers only the direct child directories as allowed directory but not the directories which a subdirictories of the direct child directory (If you understand what I mean). I tend to think so because now I have added following line `"/media/" .. os.getenv("ME") .. "/WindowsD/vaults/*"` to `auto_session_allowed_dirs` and I always open direct child directory (for example `../vaults/CheatSheets/`) with nvim and `auto-session` works like a charm.

**Reply by cameronr** [reply, marked as answer]
ah, yes. as it is currently, it expects exact directory matches (rather than doing a substring match) so if you have `/some/dir/*` it'll match `/some/dir/proj` but not `/some/dir/proj/subdir`

Are you still getting an error when restoring your session or did that resolve itself?

**Reply by MijikHna** [reply]
Thank you @cameronr. Just one last question. Do you know if this point is written down in the documentation and I just haven't properly read the docs or this edge case ist not really in the documentation?

**Reply by cameronr** [reply]
It's not addressed specifically in the docs. I'm working on a config and documentation update so will add to that. thanks!

**Reply by cameronr** [reply]
Docs are now updated
